### ISSUE TITLE
Remove "import thread" from twisted/enterprise

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8421]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)| @rodrigc reported|
|-|-|
|Trac ID|trac#8421|
|Type|enhancement|
|Created|2016-06-04 16:45:10Z|
|Branch|https://github.com/twisted/twisted/tree/8421-rodrigc-thread|

```
import thread
```
is deprecated in Python2.  Trying to **import thread** in Python3 will result in an ImportError.

The deprecation is documented here:
https://www.python.org/dev/peps/pep-3108/#thread-deprecation

and users are encouraged to not use this module.

<details><summary>Searchable metadata</summary>

```
trac-id__8421 8421
type__enhancement enhancement
reporter__rodrigc rodrigc
priority__normal normal
milestone__Python_3_x Python-3.x
branch__8421_rodrigc_thread 8421-rodrigc-thread
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1465058710630809 1465058710630809
changetime__1466964294904784 1466964294904784
version__None None
owner__rodrigc rodrigc

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_1"></a>|@rodrigc commented|
|-|-|

See:
https://github.com/twisted/twisted/pull/74

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban set owner to @rodrigc|
|-|-|

Looks good. Happy to see that we use the thread is advertised by Twisted thread code.

Waiting for test results... and to see if this change is covered.

I am not sure if the id from twisted is the same one from python.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban commented|
|-|-|

I forgot to follow up with this.

I don't know what to say about this.

The code has coverage but there is no test for it.

How can I tell that the new code is returning the same thing?

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_5"></a>|@rodrigc set owner to @adiroiban|
|-|-|

It does the right thing.  If you do something like this it works:

```
import unittest

class TestThreads(unittest.TestCase):
    def test_threadid(self):
        from twisted.python import threadable
        from twisted.python.compat import _PY3

        if _PY3:
            from _thread import get_ident
        else:
            from thread import get_ident

        self.assertEqual(threadable.getThreadID(), get_ident())

if __name__ == "__main__":
    unittest.main()

```

**thread** and **_thread** are both deprecated modules though

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban removed owner|
|-|-|

Please consider adding that test in the twisted.python.test_threadable as the current getThreadID method has no functional coverage.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_7"></a>|@rodrigc commented|
|-|-|

**thread** and **_thread** are both deprecated modules, so I am not going to add that test.
I just showed it to you as an illustration, since you asked.

under the covers, getThreadID() is calling from the Python library:

```
threading.currentThread().ident
```

How do you unittest that?  Getting the current thread id is not necessarily going to return the same value, since the test can be running under different threads.

Similarly, how would you write a unit test for:

```
os.getpid()
```
?

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)|@rodrigc set owner to @adiroiban|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_9"></a>|@rodrigc commented|
|-|-|

See new patch:

https://github.com/twisted/twisted/pull/220

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_10"></a>|@rodrigc commented|
|-|-|

Adi, can I commmit this patch?
getThreadID() already has coverage, see my comments in: [ticket:8502]

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_11"></a>|@adiroiban set owner to @rodrigc|
|-|-|

As commented in [#8502](https://github.com/twisted/twisted/issues/8502) I would argue that getThreadID has no coverage.

in the same time, this should not block this ticket.

Thanks for creating a follow up ticket and for pushing for this ticket.

Changes are ok. Please merge. Thanks again and sorry for the trouble!

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|GitHub set status to `closed`|
|-|-|

In changeset 506ee225c4d2164586678a597e7ac340f2b6e872
```
#!CommitTicketReference repository="" revision="506ee225c4d2164586678a597e7ac340f2b6e872"
Merge 8421-rodrigc-thread: Replace thread.get_ident() with twisted.python.threadable.getThreadID()

Author: rodrigc
Reviewer: adiroiban
Fixes: #8421
```
